Kernel/Arch/aarch64/PCI/Controller/BCM2712HostController.cpp

```diff
@@ -0,0 +1,263 @@
+{
+    auto domain = TRY(determine_pci_domain_for_devicetree_node(device.node(), device.node_name()));
+    auto registers_resource = TRY(device.get_resource(0));
+
+    if (registers_resource.size < sizeof(Registers))
+        return ERANGE;
+
+    auto registers = TRY(Memory::map_typed_writable<Registers volatile>(registers_resource.paddr));
+
+    // Deassert PERST# to initialize the link.
+    registers->control |= Registers::CONTROL_PERST_N;
+
+    microseconds_delay(100'000);
+
+    // Check the link state.
+    if ((registers->state & Registers::STATE_LINK_UP) != Registers::STATE_LINK_UP) {
+        dbgln("{}: Link down", device.node_name());
+
+        // We failed to initialize the link; assert PERST# again.
+        registers->control &= ~Registers::CONTROL_PERST_N;
+        return EIO;
+    }
+
+    dbgln("{}: Link up", device.node_name());
+
+    auto const* parent_node = device.node().parent();
+    if (parent_node == nullptr)
+        return EINVAL;
+
+    auto ranges = TRY(device.node().ranges());
+    for (size_t i = 0; i < ranges.entry_count(); i++) {
+        auto range = MUST(ranges.entry(i));
+
+        auto raw_pci_address = range.child_bus_address();
+        auto cpu_base = TRY(TRY(parent_node->translate_child_bus_address_to_root_address(range.parent_bus_address())).as_flatptr());
+        auto range_size = TRY(range.length().as_size_t());
+
+        if (raw_pci_address.raw().size() != sizeof(OpenFirmwareAddress))
+            return EINVAL;
+
+        OpenFirmwareAddress pci_address {};
+        memcpy(&pci_address, raw_pci_address.raw().data(), sizeof(OpenFirmwareAddress));
+
+        // FIXME: Configure the 64-bit window.
+        if (pci_address.space_type != OpenFirmwareAddress::SpaceType::Memory32BitSpace)
+            continue;
+
+        auto bus_base = pci_address.io_or_memory_space_address;
+        auto cpu_limit = cpu_base + range_size;
+
+        // Configure the CPU -> PCIe window. This determines how CPU addresses are mapped to PCIe addresses.
+
+        registers->bus_window_base_low = bus_base & 0xffff'ffff;
+        registers->bus_window_base_high = bus_base >> 32;
+
+        registers->cpu_window_base_low_and_cpu_window_limit_low = (cpu_limit & 0xfff0'0000) | ((cpu_base & 0xfff0'0000) >> 16);
+        registers->cpu_window_base_high = cpu_base >> 32;
+        registers->cpu_window_limit_high = cpu_limit >> 32;
+
+        // FIXME: Are we guaranteed to only have one 32-bit window?
+        break;
+    }
+
+    return adopt_own_if_nonnull(new (nothrow) BCM2712HostController(domain, move(registers))).release_nonnull();
+}
+
+BCM2712HostController::BCM2712HostController(PCI::Domain const& domain, Memory::TypedMapping<Registers volatile> registers)
+    : HostController(domain)
+    , m_registers(move(registers))
+{
+}
+
+ErrorOr<VirtualAddress> BCM2712HostController::get_config_window(BusNumber bus, DeviceNumber device, FunctionNumber function)
```

Comment:
Maybe more of an `activate_config_window`?

Reply:
Maybe `map_config_window` or `move_config_window` (the latter is not really fitting for b/d/f 0:0:0 though)?